common/src/main/scala/org/neo4j/spark/converter/DataConverter.scala

```diff
@@ -0,0 +1,181 @@
+        def toMap(struct: GenericRowWithSchema): Value = {
+          Values.value(
+            struct.schema.fields.map(
+              f => f.name -> convert(struct.getAs(f.name), f.dataType)
+            ).toMap.asJava)
+        }
+
+        try {
+          struct.getAs[UTF8String]("type").toString match {
+            case SchemaService.POINT_TYPE_2D => Values.point(struct.getAs[Number]("srid").intValue(),
+              struct.getAs[Number]("x").doubleValue(),
+              struct.getAs[Number]("y").doubleValue())
+            case SchemaService.POINT_TYPE_3D => Values.point(struct.getAs[Number]("srid").intValue(),
+              struct.getAs[Number]("x").doubleValue(),
+              struct.getAs[Number]("y").doubleValue(),
+              struct.getAs[Number]("z").doubleValue())
+            case SchemaService.DURATION_TYPE => Values.isoDuration(struct.getAs[Number]("months").longValue(),
+              struct.getAs[Number]("days").longValue(),
+              struct.getAs[Number]("seconds").longValue(),
+              struct.getAs[Number]("nanoseconds").intValue())
+            case SchemaService.TIME_TYPE_OFFSET => Values.value(OffsetTime.parse(struct.getAs[UTF8String]("value").toString))
+            case SchemaService.TIME_TYPE_LOCAL => Values.value(LocalTime.parse(struct.getAs[UTF8String]("value").toString))
+            case _ => toMap(struct)
+          }
+        } catch {
+          case _: Throwable => toMap(struct)
+        }
+      }
+      case unsafeArray: UnsafeArrayData => {
+        val sparkType = dataType match {
+          case arrayType: ArrayType => arrayType.elementType
+          case _ => dataType
+        }
+        val javaList = unsafeArray.toSeq[AnyRef](sparkType)
+          .map(elem => convert(elem, sparkType))
+          .asJava
+        Values.value(javaList)
+      }
+      case unsafeMapData: UnsafeMapData => { // Neo4j only supports Map[String, AnyRef]
+        val mapType = dataType.asInstanceOf[MapType]
+        val map: Map[String, AnyRef] = (0 until unsafeMapData.numElements())
+          .map(i => (unsafeMapData.keyArray().getUTF8String(i).toString, unsafeMapData.valueArray().get(i, mapType.valueType)))
+          .toMap[String, AnyRef]
+          .mapValues(innerValue => convert(innerValue, mapType.valueType))
+          .toMap[String, AnyRef]
+        Values.value(map.asJava)
+      }
+      case string: UTF8String => convert(string.toString)
+      case _ => Values.value(value)
+    }
+  }
+}
+
+object Neo4jToSparkDataConverter {
+  def apply(): Neo4jToSparkDataConverter = new Neo4jToSparkDataConverter()
+}
+
+class Neo4jToSparkDataConverter extends DataConverter[Any] {
+  override def convert(value: Any, dataType: DataType): Any = {
+    if (dataType != null && dataType == DataTypes.StringType && value != null && !value.isInstanceOf[String]) {
+      convert(Neo4jUtil.mapper.writeValueAsString(value), dataType)
+    } else {
+      value match {
+        case node: InternalNode => {
```

Comment:
do we need the concrete internal types here?
ain't the type outlined in https://neo4j.com/docs/java-reference/current/extending-neo4j/values-and-types/ enough?

Reply:
changed here, in `InternalRelationship -> Relationship` and `InternalIsoDuration -> IsoDuration`